corehq/mobile_ucr_v2_update_script.py

```diff
@@ -0,0 +1,114 @@
+
+
+import json
+import re
+import traceback
+
+from corehq.apps.app_manager.dbaccessors import (
+    get_latest_app_ids_and_versions,
+    get_apps_by_id,
+)
+from corehq.apps.app_manager.const import MOBILE_UCR_VERSION_2
+from corehq.toggles import MOBILE_UCR
+from corehq.toggles.shortcuts import find_domains_with_toggle_enabled
+from corehq.util.log import with_progress_bar
+
+
+def save_in_log(file, data):
+    print(data)
+    file.write(data + '\n')
+
+
+def save_as_ndjson(path, data):
+    with open(path, 'a') as file:
+        print(json.dumps(data, separators=(',', ':')), file=file)
+
+
+def read_ndjson_file(path):
+    with open(path, 'r') as file:
+        return set(json.loads(line) for line in file.readlines())
+
+
+def has_non_v2_form(domain, app, log_file):
+    for form in app.get_forms():
+        save_in_log(log_file, f"Processing Form: {domain}: {form.name}")
+        # The second condition should always be False if the first one is
+        # but just as a precaution we check for it
+        if V1_FIXTURE_IDENTIFIER in form.source or RE_V1_ALL_REFERENCES.search(form.source):
+            save_in_log(log_file, f"App Contains V1 Refs: {domain}: {app.name}")
+            return True
+    return False
+
+
+def update_app(domain, app, log_file):
+    save_in_log(log_file, f"Updating App: {domain}: {app.name}: {app.id}")
+    app.mobile_ucr_restore_version = MOBILE_UCR_VERSION_2
+    app.save()
+
+
+PROCESSED_DOMAINS_PATH = '/home/zandre/cchq/updated_domains.ndjson'
+LOG_FILE = '/home/zandre/cchq/update_to_v2_ucr_script.log'
+
+V1_FIXTURE_IDENTIFIER = 'src="jr://fixture/commcare:reports'
+V1_FIXTURE_PATTERN = r'<.*src="jr://fixture/commcare:reports.*>'
+V1_REFERENCES_PATTERN = r"<.*instance\('reports'\)/reports/.*>"
+RE_V1_ALL_REFERENCES = re.compile(f"{V1_FIXTURE_PATTERN}|{V1_REFERENCES_PATTERN}")
+
+
+skip_domains = set()
+
+
+def process():
+    try:
+        processed_domains = read_ndjson_file(PROCESSED_DOMAINS_PATH)
+    except FileNotFoundError:
+        processed_domains = set()
+
+    mobile_ucr_domains = find_domains_with_toggle_enabled(MOBILE_UCR)
+
+    log_file = open(LOG_FILE, 'a')
+
+    save_in_log(log_file, f"Number of domains with mobile ucr flag enabled: {len(mobile_ucr_domains)} ")
+
+    for domain in with_progress_bar(mobile_ucr_domains):
+        if domain in processed_domains:
+            save_in_log(log_file, f"Already processed domain: {domain}")
+            continue
+        if domain in skip_domains:
+            save_in_log(log_file, f"Skipped domain: {domain}")
+            continue
+
+        save_in_log(log_file, f"Processing domain: {domain} ...")
+        app_ids = list(get_latest_app_ids_and_versions(domain))
+        apps = get_apps_by_id(domain, app_ids)
+        for app in apps:
+            try:
+                # Don't look at app.is_released since the latest version might not be released yet
```

Comment:
Confirming that we are just picking app and not releases.
I believe one way to confirm that is by checking `copy_of` property on the application. Is it blank for applications and not blank for releases. it would be a good check to have though I assume you are anyway only picking up applications and not releases.

Reply:
> Confirming that we are just picking app and not releases.

Could you please elaborate on what you mean by app vs release? Are you referring to app versions that have been marked as "Released" or are you referring to the app versions themselves?

This comment refers that we are picking the latest version of each app and we don't want to consider `app.is_released` since the latest version of an app might still be set to "Test" and not "Released"

Reply:
> Are you referring to app versions that have been marked as "Released" or are you referring to the app versions themselves?

app versions or app builds.

Reply:
We are getting the applications themselves using `get_apps_by_id()` which returns a list of `Application` objects. Since we are only getting the latest release of each app, these retrieved application objects should be the current build of the app.

@mkangia I'm not sure if the above answers your question, but let me know if it doesn't and I'd be happy to offline on this.

Reply:
> Since we are only getting the latest release of each app, these retrieved application objects should be the current build of the app.

I think there is a misunderstanding of how apps work.

So, an application is actually a running doc and the versions are saved instances of that application at different points(versions). My understanding is that here we want to update the application and not the latest created version/build.

Seems like we are okay here since you are using `get_latest_app_ids_and_versions` which is using app_manager/applications_brief couch view which only keeps applications and not builds since it only saves docs that have [copy_of set as null](https://github.com/dimagi/commcare-hq/blob/73d612ec2bd6dce6572dde770dfbde084495369d/corehq/apps/app_manager/_design/views/applications_brief/map.js#L2)